tests/kernel/mem_protect/userspace/src/main.c

```diff
@@ -488,6 +516,167 @@ static void read_kobject_user_pipe(void)
+	 * Cannot add a thread to two memory domains
+	 */
+	k_mem_domain_add_thread(&app1_domain, k_current_get());
+
+		/* These lines of code should not be reached */
+	zassert_unreachable("Adding a thread to multiple memory domains"
+			" did not fault\n");
+}
+
+/* Verify that a thread can belong to a single memory domain
+ * at any point of time
+ * Try to add a thread to two memory domains at a time
+ */
+static void thread_in_multiple_domains(void)
+{
+	k_mem_domain_init(&app0_domain, ARRAY_SIZE(app0_parts), app0_parts);
+	k_mem_domain_init(&app1_domain, ARRAY_SIZE(app1_parts), app1_parts);
+
+	k_tid_t thread = k_thread_create(&kthread_thread, kthread_stack,
+			STACKSIZE, (k_thread_entry_t)kthread_entry, NULL, NULL,
+			NULL, K_PRIO_PREEMPT(1), 0, K_NO_WAIT);
+
+	k_sem_take(&sync_sem, SEM_TIMEOUT);
+
+	zassert_true(thread->mem_domain_info.mem_domain == &app0_domain,
+			"Adding a thread to 2 memory domains didnot fault");
+}
+
+/**************************************************************************/
+/* Test to validate that Threads in the same memory domain have
+ * the same access permissions to the memory partitions
+ * belonging to the memory domain
+ */
+struct k_mem_domain app_domain;
+u8_t MEM_DOMAIN_ALIGNMENT app_buf[MEM_REGION_ALLOC];
+K_MEM_PARTITION_DEFINE(app_part, app_buf, sizeof(app_buf),
+		K_MEM_PARTITION_P_RO_U_RO);
+struct k_mem_partition *app_parts[] = {
+	    &app_part
+};
+
+static void kentry(void)
+{
+	printk("Read the partition app_buf[0]=%d\n", app_buf[0]);
+
+	/* Write to the partition, this should generate fault */
+	app_buf[0] = 10;
+
+	/* These lines of code should not be reached */
+	zassert_unreachable("Both the threads should have same "
+			"permissions to the domain which they belong to\n");
+}
+
+static void check_access_perms(void)
+{
+	/* The following should cause a fault */
+	expect_fault = true;
+	expected_reason = REASON_HW_EXCEPTION;
+	BARRIER();
+
+	/* Initialize the memory domain */
+	k_mem_domain_init(&app_domain, ARRAY_SIZE(app_parts),
+			app_parts);
+
+	k_mem_domain_add_thread(&app_domain, k_current_get());
+
+	/* Create a user thread */
+	k_thread_create(&utid, ustack, MEM_DOMAIN_STACK_SIZE,
+			(k_thread_entry_t)kentry, NULL, NULL, NULL,
+			10, K_USER, K_NO_WAIT);
+
+	k_sem_take(&sync_sem, SEM_TIMEOUT);
+}
+
+/***********************************************************************/
+/* Validate k_mem_domain_destroy API */
+
+static void destroy_mem_domain(void)
+{
+	k_mem_domain_init(&app0_domain, ARRAY_SIZE(app0_parts), app0_parts);
+
+	k_mem_domain_add_thread(&app0_domain, k_current_get());
+
+	if (_current->mem_domain_info.mem_domain == &app0_domain) {
+		k_mem_domain_destroy(&app0_domain);
+		zassert_true(_current->mem_domain_info.mem_domain !=
+				&app0_domain, "The thread has reference to"
+				" memory domain which is already destroyed");
+	} else {
+		zassert_unreachable("k_mem_domain_add_thread() failed\n");
+	}
+}
+
```

Comment:
The memory domain tests all appear to fail on ARM w/ current master (which has pr-4974 merged), with:
ASSERTION FAIL [writable != executable] @ /home/sds/zephyr-project/kernel/mem_domain.c:21:
        writable page not executable

Also I still have doubts about whether these belong here, particularly the ones that don't create any userspace thread.

Reply:
These tests are generic, I don't think the failure is because of the test case.  I will merge this test case with #5197 once it is merged. It is more related to those set of test cases.